typed-racket-lib/typed-racket/typecheck/tc-structs.rkt

```diff
@@ -538,3 +543,18 @@
                           (list #'fld ...)
                           (list ty ...)
                           opts.kernel-maker)]))
+
+(define/cond-contract (tc/make-struct-type-property prop-name pred-id struct-property-ty)
+  (syntax? syntax? Type? . c:-> . tc-results/c)
+  (match struct-property-ty
+    [(Struct-Property: ty #f)
+     (set-struct-property-pred! struct-property-ty pred-id)
+     (define has-spty (make-Has-Struct-Property prop-name))
+     (values->tc-results (make-Values (list (-result struct-property-ty -true-propset -empty-obj)
+                                            (-result (make-pred-ty has-spty))
+                                            (-result (make-Fun (list (-Arrow (list has-spty) ty
+                                                                             #:props (-PS (-is-type 0 -Self)
+                                                                                          (-not-type 0 -Self))))))))
+                                             '())]
+    [(Struct-Property: ty _) (tc-error "the predicate for ~a should not be set mannualy " (syntax-e prop-name))]
```

Comment:
typo, want "manually"

it'd also be good to add "use #f instead" if that's what the user should do